espaloma/data/qcarchive_utils.py

```diff
@@ -102,21 +136,83 @@ def get_graph(collection, record_name):
     return g
 
 
-def fetch_td_record(record: ptl.models.torsiondrive.TorsionDriveRecord):
-    final_molecules = record.get_final_molecules()
-    final_results = record.get_final_results()
+def get_graph(collection, record_name, spec_name="default"):
+    """
+    Processes the qcportal data for a given record name.
+
+    Parameters
+    ----------
+    collection, qcportal dataset, required
+        The instance of the qcportal dataset
+    record_name, str, required
+        The name of a give record
+    spec_name, str, default="default"
+        Retrieve data for a given qcportal specification.
+    Returns
+    -------
+        Graph
+    """
+    # get record and trajectory
+    record = collection.get_record(record_name, specification_name=spec_name)
+    entry = collection.get_entry(record_name)
+
+    g = process_record(record, entry)
+
+    return g
+
 
-    angle_keys = list(final_molecules.keys())
+def get_graphs(collection, record_names, spec_name="default"):
+    """
+    Processes the qcportal data for a given set of record names.
+    This uses the qcportal iteration functions which are faster than processing
+    records one at a time
+
+    Parameters
+    ----------
+    collection, qcportal dataset, required
+        The instance of the qcportal dataset
+    record_name, str, required
+        The name of a give record
+    spec_name, str, default="default"
+        Retrieve data for a given qcportal specification.
+    Returns
+    -------
+    list(graph)
+        Returns a list of the corresponding graph for each record name
+    """
+    g_list = []
+    for record, entry in zip(
+            collection.iterate_records(record_names, specification_names=[spec_name]),
+            collection.iterate_entries(record_names),
+    ):
+        g = process_record(record, entry)
+        g_list.append(g)
+
+    return g_list
+
+
+def fetch_td_record(record: qcportal.torsiondrive.record_models.TorsiondriveRecord):
+    molecule_optimization = record.optimizations
+
+    angle_keys = list(molecule_optimization.keys())
 
     xyzs = []
     energies = []
     gradients = []
 
     for angle in angle_keys:
-        result = final_results[angle]
-        mol = final_molecules[angle]
+        # NOTE: this is calling the first index of the optimization array
+        # this gives the same value as the prior implementation, but I wonder if it
+        # should be molecule_optimization[angle][-1] in both cases
```

Comment:
@kntkb or @yuanqing-wang thoughts?

Reply:
So I've been trying to figure out the structure of the torsion drive datasets (since I have not looked at them really yet prior to this).    Considering the example dataset I used in test (I'll put the code below), each angle has n-number of unique initial conformations that are then optimized.  In this case, there are 4 configurations (each that has their own trajectory).   So I suppose choosing the first vs the last is somewhat irrelevant (I was initially thinking this was a set of chained optimizations, hence my comment...don't ask why I was thinking that). 

Should each of these conformations be considered and added to the datasets rather than just arbitrarily picking one?


```
from espaloma.data import qcarchive_utils
import numpy as np

record_name = "[h]c1c(c(c(c([c:1]1[n:2]([c:3](=[o:4])c(=c([h])[h])[h])c([h])([h])[h])[h])[h])n(=o)=o)[h]"
name = "OpenFF Amide Torsion Set v1.0"
collection_type = "torsiondrive"
collection, record_names = qcarchive_utils.get_collection(qcarchive_utils.get_client(), collection_type, name)
record_info = collection.get_record(record_name, specification_name="default")

molecule_optimization = record_info.optimizations
angle_keys = list(molecule_optimization.keys())

angle = angle_keys[0]
mol = molecule_optimization[angle][0].final_molecule
result = molecule_optimization[angle][0].trajectory[-1].properties
```

looking at the actual configurations:

```
for i in range(len(molecule_optimization[angle])):
    init =  molecule_optimization[angle][i].initial_molecule.geometry
    final = molecule_optimization[angle][i].final_molecule.geometry
    print(init,"\n-\n", final, "\n--\n")
```

Reply:
> @kntkb or @yuanqing-wang thoughts?

I don't know off the top of my head, but I've played around with different QCArchive workflows in the past. I may have some notes left somewhere, so I'll catch up shortly (tomorrow?).

Reply:
Oh, the api and the way you access the data changed using qcportal v0.5...

Reply:
I think in the older version of qcportal, "get_final_molecule()" just picked the first one in the array.  The full array was still part of the data record, just you had to dig through the qcvars or something to access. From conversations with Ben, there was a lot of trying to force records into a very rigid schema in the old version; he opted to break the schema in a lot of cases to just make it easier to access the relevant information (and make it clearer what information is available). 

As I mentioned in an early comment, it seems that for each angle, multiple (in this case 4) independent starting configurations were used.  It seems like it would be better to have the code return data for each replicate, but I'm not sure how this would impact any workflows that use this function.